ui-library-filesharing-ui-components/app/src/ChatComponents.tsx

```diff
@@ -0,0 +1,137 @@
+  const [files, setFiles] = React.useState<File[] | []>();
+  const [activeFileUploads, setActiveFileUploads] = React.useState<
+    ActiveFileUpload[] | []
+  >([]);
+
+  const updateFileUploadProgress = (
+    fileId: string,
+    progress: number,
+    complete: boolean = false
+  ) => {
+    const updatedData = allActiveFileUploads.current.map((active) => {
+      if (active.id === fileId) {
+        return {
+          ...active,
+          progress,
+          uploadComplete: complete,
+        };
+      }
+      return active;
+    });
+    allActiveFileUploads.current = updatedData;
+    setActiveFileUploads(updatedData);
+  };
+
+  const uploadFile = async (file: File): Promise<void> => {
+    const extension = file.name.split(".").pop() || "";
+    const uniqueFileName = `${v4()}-${file.name}`;
+    const data = new Form();
+    data.append("file", file);
+
+    axios
+      .request({
+        method: "post",
+        url: `/api/UploadFileToAzureBlobStore?filename=${uniqueFileName}`,
+        data: data,
+        onUploadProgress: (p) => {
+          updateFileUploadProgress(file.name, p.loaded / p.total);
+        },
+      })
+      .then((res) => {
+        updateFileUploadProgress(file.name, 1, true);
+        completedFileUploads.current = [
+          ...completedFileUploads.current,
+          {
+            name: file.name,
+            extension,
+            url: res.data.url,
+          },
+        ];
+      })
+      .catch((err) => {
+        console.log(err);
+      });
+  };
+
+  React.useEffect(() => {
+    if (files?.length) {
+      files.forEach((file) => {
+        uploadFile(file);
+      });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [files]);
+
+  const onChange = (files: FileList | null) => {
+    if (!files) return;
+    setFiles(Array.from(files));
+    allActiveFileUploads.current = Array.from(files).map((file) => ({
+      filename: file.name,
+      id: file.name,
+      progress: 0,
+    }));
+    setActiveFileUploads(allActiveFileUploads.current);
+  };
+
+  const messageThreadProps = usePropsFor(MessageThread);
+  const sendBoxProps = usePropsFor(SendBox);
+
+  return (
+    <div style={{ height: "100%", width: "100%" }}>
+      {messageThreadProps && <MessageThread {...messageThreadProps} />}
+      {sendBoxProps && (
+        <SendBox
+          {...sendBoxProps}
+          activeFileUploads={activeFileUploads}
+          onCancelFileUpload={(id: string) => {
+            allActiveFileUploads.current = allActiveFileUploads.current.filter(
+              (upload) => upload.id !== id
+            );
+            setActiveFileUploads(allActiveFileUploads.current);
+          }}
+          onSendMessage={async (message: string) => {
+            sendBoxProps.onSendMessage(message, {
+              metadata: {
+                fileSharingMetadata: JSON.stringify(
+                  completedFileUploads.current
```

Comment:
Will this attach the right metadata?

Updating `completedFileUploads` does not trigger a render. So the also value of `completedFileUploads` could be stale here. I think this works because `activeFileUploads` is _always_ updated along with `completedFileUploads`, which causes a render.
Should document that.

Reply:
The component doesn't need to re-render for the value of `completedFileUploads` to be up to date. Ref preserves value between renders and always has the latest value.
We update the value of `completedFileUploads` whenever an upload is completed.
Tested this and it works as expected.

Reply:
My bad. That's the whole point of a ref.